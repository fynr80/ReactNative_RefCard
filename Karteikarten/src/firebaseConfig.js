
import firebase from 'firebase/compat/app';

const firebaseConfig = {
    apiKey: "AIzaSyBksswFLr1YB6en-sRf8mqSFle7bhAzSgU",
    authDomain: "swtp-812df.firebaseapp.com",
    databaseURL: "https://swtp-812df-default-rtdb.firebaseio.com",
    projectId: "swtp-812df",
    storageBucket: "swtp-812df.appspot.com",
    messagingSenderId: "44236558384",
    appId: "1:44236558384:web:2f15f6b5fb190634b71df5",
    measurementId: "G-W33L0G7D18"
};
firebase.initializeApp(firebaseConfig);
export {firebase};
