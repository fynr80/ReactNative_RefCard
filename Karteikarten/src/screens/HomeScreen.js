/**
 * @file HomeScreen of RefCard app
 * @author Fatih Yanar <fatih.yanar@mni.thm.de>
 */


import * as React from 'react';
import {
    View,
    Text,
    StyleSheet,
    TouchableOpacity,
    FlatList,
    Image}
from 'react-native';
import {useEffect,useState} from "react";

import {SafeAreaView} from 'react-native-safe-area-context';

//import {LinearGradient} from "expo-linear-gradient";

import SetCategories from "../components/SetCategories";

import firebase from 'firebase/compat/app';

export default function HomeScreen({ }) {

    const [buttonFlag, setButtonFlag] = useState(true);
    const [myCategoriesArr, setMyCategoriesArr] = useState([]);


    useEffect(() => {
        getCategories();
    }, [])

    const setButtonFlag_func = () => {
        if (!buttonFlag) setButtonFlag(true)
    }

    const setButtonFlag_func2 = () => {
        if (buttonFlag) setButtonFlag(false)
    }

    /**
     * Get all Categories and save them into the Array to be displayed
     */
    const getCategories = () => {
        firebase.firestore()
            .collection('category')
            .onSnapshot(snapshot => {
            const arr = snapshot.docs.map((doc) => ({
                id: doc.id,
                name: doc.data().titel,
                categoryColor:doc.data().color
            }))
            setMyCategoriesArr(arr)
        })
    }


    /**
     * Toggles one of the two Buttons when Pressed
     * @param buttonFlag : Boolean
     * @returns Style of the Button
     */
    const btnStyle = function (buttonFlag) {
        if (buttonFlag) {
            return [styles.twoButtonFlagTrue]
        }
        return [styles.twoButtonFlagFalse]
    }


    return (
        <SafeAreaView style={styles.container}>
            <View style={styles.header}>
                <Image
                    style={{height:90,width:120,backgroundColor:'white'}}
                    source={require('../asserts/Logonew.png')}
                />
            </View>
            <View style={styles.lastOpenedTextContainer}>
                <Text style={styles.lastOpenedText} testID={"LastOpenedText"}>Last Opened</Text>
            </View>
            <TouchableOpacity  style={styles.lastOpenedItems}>
                <View style={styles.lastItem}>
                {/*<LinearGradient style={styles.lastItem} colors={['#2B32B2','#1488CC']}>*/}
                    <Text style={[styles.lastOpenedText,{color: 'white',paddingLeft:10,}]}>Diskrete Mathematik</Text>
                    <Text style={[styles.lastOpenedText,{color: 'white',paddingLeft:10,fontSize: 15,fontWeight: 'normal',opacity:0.8}]}>15 Questions</Text>
                {/*</LinearGradient>*/}
                </View>
            </TouchableOpacity>
            <View style={styles.buttons}>
                <TouchableOpacity testID={"MyCatBtn"} onPress={setButtonFlag_func}
                                  style={btnStyle(buttonFlag)}>
                    <Text testID={"MyCatText"} style={[styles.lastOpenedText,{color: buttonFlag ? '#F6F6F6': 'black',textAlign:'center'}]}>My Category</Text>
                </TouchableOpacity>
                <TouchableOpacity  testID={"AllCatBtn"} onPress={setButtonFlag_func2}
                                   style={btnStyle(!buttonFlag)}>
                    <Text testID={"AllCatText"} style={[styles.lastOpenedText,{color: buttonFlag ? 'black': '#F6F6F6',  textAlign:'center'}]}>All Category</Text>
                </TouchableOpacity>
            </View>
            <View style={styles.catList}>
                <View style={styles.categoryItem}>
                    <FlatList
                        testID={"ItemsList"}
                        numColumns={1}
                        data={myCategoriesArr}
                        keyExtractor={(item, index) => item + index}
                        renderItem={({item}) => <SetCategories name={item.name} id={item.id} categoryColor={item.categoryColor}/>}
                    />
                </View>
            </View>
        </SafeAreaView>
    );
}

const styles = StyleSheet.create({
    container: {
        flex: 1,
        justifyContent: 'center',
        alignContent:'flex-start',
        backgroundColor:'white',
    },
    header:{
        backgroundColor:'white',
        flexDirection:'row',
        flex:1.9,
        alignItems:'center',
        paddingLeft:2,
        paddingRight:2,
        justifyContent:'center',
        marginTop:3
    },
    headerText:{
        fontSize:20,
        fontWeight:'bold',
        marginLeft:90
    },
    lastOpenedTextContainer: {
        backgroundColor:'white',
        flex:1,
        alignItems:'center',
        justifyContent:'flex-start',
        flexDirection:'row',
        paddingLeft:10
    },
    lastOpenedText: {
        fontSize:20,
        fontWeight:'bold',
        color:'black'
    },
    lastOpenedItems: {
        backgroundColor:'white',
        flexDirection:'row',
        flex:2,
        alignItems:'center',
        justifyContent:'space-between',
        paddingLeft:10,
        paddingRight:10,
        width:'100%'
    },
    lastItem: {
        justifyContent:'center',
        width:'100%',
        height:'80%',
        backgroundColor:'blue',
        borderRadius:10,
    },
    buttons:{
        backgroundColor:'white',
        flexDirection:'row',
        flex:2,
        alignItems:'center',
        justifyContent:'space-between',
        paddingLeft:10,
        paddingRight:10
    },
    twoButtonFlagTrue: {
        justifyContent:'center',
        width:'46%',
        height:'60%',
        backgroundColor:'#FD4365',
        borderRadius:30,
        shadowColor: "#000",
        shadowOffset: {
            width: 0,
            height: 7,
        },
        shadowOpacity: 0.43,
        shadowRadius: 9.51,
        elevation: 15,
    },
    twoButtonFlagFalse: {
        justifyContent:'center',
        width:'46%',
        height:'60%',
        backgroundColor:'#FBF9F9',
        borderRadius:30,
        shadowColor: "#000",
        shadowOffset: {
            width: 0,
            height: 7,
        },
        shadowOpacity: 0.43,
        shadowRadius: 9.51,
        elevation: 15,
    },
    catList:{
        backgroundColor:'white',
        flexDirection:'column',
        flex:8,
        alignItems:'center',
        justifyContent:'space-between',
        paddingLeft:10,
        paddingRight:10,
    },
    categoryItem :{
        backgroundColor:'white',
    }
});
