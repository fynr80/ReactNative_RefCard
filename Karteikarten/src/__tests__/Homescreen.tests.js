import React from 'react';
import renderer from 'react-test-renderer';
import HomeScreen from "../screens/HomeScreen";
import {fireEvent, render, screen} from "@testing-library/react-native";
import firebaseConfig from "../firebaseConfig";
import {toMatchImageSnapshot} from "jest-image-snapshot"

const puppeteer = require("puppeteer")

expect.extend({toMatchImageSnapshot});

describe("HomeScreen Component", () => {
    let browser;
    jest.setTimeout(35000);

    beforeAll(async () => {
        browser = await puppeteer.launch({args: ['--no-sandbox', '--disable-setuid-sandbox']});
    })

    it('renders correctly', async () => {
        const tree = renderer.create(<HomeScreen/>).toJSON();
        expect(tree).toMatchSnapshot();
    });

    it('renders Text correctly', async () => {
        render(<HomeScreen/>);
        const text = screen.getByTestId("LastOpenedText");
        expect(text.props.children).toEqual("Last Opened")
    });

    it("it makes a screenshot", async () => {
        const page = await browser.newPage();
        await page.goto('http://localhost:19006');

        await new Promise((r) => setTimeout(r, 5000));

        const image1 = await page.screenshot();

        const gotoStatisticBtnSelector = "#root > div > div > div > div > div > div.css-view-1dbjc4n.r-flex-13awgt0 > div > div.css-view-1dbjc4n.r-bottom-1p0dtai.r-display-6koalj.r-left-1d2f490.r-pointerEvents-12vffkv.r-position-u8s1d.r-right-zchlnj.r-top-ipm5af > div.css-view-1dbjc4n.r-flex-13awgt0.r-pointerEvents-12vffkv > div > div > div > div.css-view-1dbjc4n.r-flex-13awgt0 > div > div.css-view-1dbjc4n.r-bottom-1p0dtai.r-left-1d2f490.r-pointerEvents-105ug2t.r-right-zchlnj > div.css-view-1dbjc4n.r-flex-13awgt0.r-flexDirection-18u37iz > a:nth-child(2)";
        await page.waitForSelector(gotoStatisticBtnSelector);
        const gotoStatisticBtn = await page.$(gotoStatisticBtnSelector);
        gotoStatisticBtn.click();

        await new Promise((r) => setTimeout(r, 5000));

        const image2 = await page.screenshot();

        expect(image1).toMatchImageSnapshot();
        expect(image2).toMatchImageSnapshot();
    })

    it('buttons should have correct styling', async () => {
        const {getByTestId} = render(<HomeScreen/>);
        const myCatBtn = getByTestId("MyCatBtn");
        const allCatBtn = getByTestId("AllCatBtn");

        expect(myCatBtn.props.style.backgroundColor).toEqual('#FD4365');
        expect(allCatBtn.props.style.backgroundColor).toEqual('#FBF9F9');
    });

    it('buttons should have different color after Press', async () => {
        const {getByTestId} = render(<HomeScreen/>);
        const myCatBtn = getByTestId("MyCatBtn");
        const allCatBtn = getByTestId("AllCatBtn");

        fireEvent.press(getByTestId("AllCatBtn"));

        expect(allCatBtn.props.style.backgroundColor).toEqual('#FD4365');
        expect(myCatBtn.props.style.backgroundColor).toEqual('#FBF9F9');
    });

    afterAll(async () => {
        await browser.close();
    })

})
