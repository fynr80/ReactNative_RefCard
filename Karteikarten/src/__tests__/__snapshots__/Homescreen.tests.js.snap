// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HomeScreen Component renders correctly 1`] = `
<RNCSafeAreaView
  style={
    Object {
      "alignContent": "flex-start",
      "backgroundColor": "white",
      "flex": 1,
      "justifyContent": "center",
    }
  }
>
  <View
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "white",
        "flex": 1.9,
        "flexDirection": "row",
        "justifyContent": "center",
        "marginTop": 3,
        "paddingLeft": 2,
        "paddingRight": 2,
      }
    }
  >
    <Image
      source={1}
      style={
        Object {
          "backgroundColor": "white",
          "height": 90,
          "width": 120,
        }
      }
    />
  </View>
  <View
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "white",
        "flex": 1,
        "flexDirection": "row",
        "justifyContent": "flex-start",
        "paddingLeft": 10,
      }
    }
  >
    <Text
      style={
        Object {
          "color": "black",
          "fontSize": 20,
          "fontWeight": "bold",
        }
      }
      testID="LastOpenedText"
    >
      Last Opened
    </Text>
  </View>
  <View
    accessible={true}
    focusable={false}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "white",
        "flex": 2,
        "flexDirection": "row",
        "justifyContent": "space-between",
        "opacity": 1,
        "paddingLeft": 10,
        "paddingRight": 10,
        "width": "100%",
      }
    }
  >
    <View
      style={
        Object {
          "backgroundColor": "blue",
          "borderRadius": 10,
          "height": "80%",
          "justifyContent": "center",
          "width": "100%",
        }
      }
    >
      <Text
        style={
          Array [
            Object {
              "color": "black",
              "fontSize": 20,
              "fontWeight": "bold",
            },
            Object {
              "color": "white",
              "paddingLeft": 10,
            },
          ]
        }
      >
        Diskrete Mathematik
      </Text>
      <Text
        style={
          Array [
            Object {
              "color": "black",
              "fontSize": 20,
              "fontWeight": "bold",
            },
            Object {
              "color": "white",
              "fontSize": 15,
              "fontWeight": "normal",
              "opacity": 0.8,
              "paddingLeft": 10,
            },
          ]
        }
      >
        15 Questions
      </Text>
    </View>
  </View>
  <View
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "white",
        "flex": 2,
        "flexDirection": "row",
        "justifyContent": "space-between",
        "paddingLeft": 10,
        "paddingRight": 10,
      }
    }
  >
    <View
      accessible={true}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "backgroundColor": "#FD4365",
          "borderRadius": 30,
          "elevation": 15,
          "height": "60%",
          "justifyContent": "center",
          "opacity": 1,
          "shadowColor": "#000",
          "shadowOffset": Object {
            "height": 7,
            "width": 0,
          },
          "shadowOpacity": 0.43,
          "shadowRadius": 9.51,
          "width": "46%",
        }
      }
      testID="MyCatBtn"
    >
      <Text
        style={
          Array [
            Object {
              "color": "black",
              "fontSize": 20,
              "fontWeight": "bold",
            },
            Object {
              "color": "#F6F6F6",
              "textAlign": "center",
            },
          ]
        }
        testID="MyCatText"
      >
        My Category
      </Text>
    </View>
    <View
      accessible={true}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "backgroundColor": "#FBF9F9",
          "borderRadius": 30,
          "elevation": 15,
          "height": "60%",
          "justifyContent": "center",
          "opacity": 1,
          "shadowColor": "#000",
          "shadowOffset": Object {
            "height": 7,
            "width": 0,
          },
          "shadowOpacity": 0.43,
          "shadowRadius": 9.51,
          "width": "46%",
        }
      }
      testID="AllCatBtn"
    >
      <Text
        style={
          Array [
            Object {
              "color": "black",
              "fontSize": 20,
              "fontWeight": "bold",
            },
            Object {
              "color": "black",
              "textAlign": "center",
            },
          ]
        }
        testID="AllCatText"
      >
        All Category
      </Text>
    </View>
  </View>
  <View
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "white",
        "flex": 8,
        "flexDirection": "column",
        "justifyContent": "space-between",
        "paddingLeft": 10,
        "paddingRight": 10,
      }
    }
  >
    <View
      style={
        Object {
          "backgroundColor": "white",
        }
      }
    >
      <RCTScrollView
        data={Array []}
        disableVirtualization={false}
        getItem={[Function]}
        getItemCount={[Function]}
        horizontal={false}
        initialNumToRender={10}
        keyExtractor={[Function]}
        maxToRenderPerBatch={10}
        onContentSizeChange={[Function]}
        onEndReachedThreshold={2}
        onLayout={[Function]}
        onMomentumScrollEnd={[Function]}
        onScroll={[Function]}
        onScrollBeginDrag={[Function]}
        onScrollEndDrag={[Function]}
        removeClippedSubviews={false}
        renderItem={[Function]}
        scrollEventThrottle={50}
        stickyHeaderIndices={Array []}
        testID="ItemsList"
        updateCellsBatchingPeriod={50}
        viewabilityConfigCallbackPairs={Array []}
        windowSize={21}
      >
        <View />
      </RCTScrollView>
    </View>
  </View>
</RNCSafeAreaView>
`;
